JBTM-3213 Allow request timeouts to be configurable. Enable throttlin…

mmusgrov opened the pull request:
https://issues.redhat.com/browse/JBTM-3213

!MAIN !CORE !QA_JTA !QA_JTS_JDKORB !QA_JTS_OPENJDKORB !QA_JTS_JACORB !BLACKTIE !XTS !PERF NO_WIN !RTS !AS_TESTS !TOMCAT !JACOCO
LRA

This is a partial fix for JBTM-3213 and the PR increases the resilience of LRA to overload. The PR:

- allows throttling of LRA start requests to be configurable;
- allows clients to control how long coordinator requests can take before giving up. The PR adds separate properties for things like starting, joining, ending LRAs etc. We need separate properties since some request failures are more serious than others (eg failure to abort an LRA is significantly more problematic than are failures to start or join with LRAs).

There is a lot we can do in this area [fault resilience] and we have a few JIRAs for tackling the various requirements.
But this PR is a basic minimum to reduce the number of errors caused by overloading the system in which LRA operates.

ochaloup commented on rts/lra/coordinator/pom.xml:
```diff
@@ -28,6 +28,11 @@
+        <dependency>
+            <groupId>org.eclipse.microprofile.fault-tolerance</groupId>
+            <artifactId>microprofile-fault-tolerance-api</artifactId>
+            <version>2.1.1</version>
```
The versions would be better to be configurable with properties.

mmusgrov replied:
Done

ochaloup commented on rts/lra/client/src/main/java/io/narayana/lra/client/NarayanaLRAClient.java:
```diff
@@ -124,7 +124,11 @@
      * WARNING: NarayanaLRAClient is an internal utility class and is subject to change
      * (but it will be made available with improved guarantees in a subsequent release).
      */
-    private static final long CLIENT_TIMEOUT = Long.getLong("lra.internal.client.timeout", 10);
```
Would not be good to have one system property to define all client timeout values at once and on top of that to provide ability to granularly change the particular value?

mmusgrov replied:
Well that would be more code than I wanted to provide at this stage and we'd have document clearly which properties take precedence over others. I will rename the properties so that we can do what you request in a future revision, something like:

```java
    private static final long CLIENT_TIMEOUT = Long.getLong("lra.internal.client.timeout", 10);
    private static final long START_TIMEOUT = Long.getLong("lra.internal.client.timeouts.tart", 10);
    private static final long JOIN_TIMEOUT = Long.getLong("lra.internal.client.timeout.join", 10);
etc
```

ochaloup replied:
ok

mmusgrov replied:
Actually the change was rather trivial so I incorporated it into the PR.

ochaloup commented on rts/lra/coordinator/src/main/java/io/narayana/lra/coordinator/api/Coordinator.java:
```diff
@@ -189,6 +191,7 @@ public LRAData getLRAInfo(
     @POST
     @Path("start")
     @Produces(MediaType.TEXT_PLAIN)
+    @Bulkhead
```
I'm interested what is the reason for taking the faultolerance `@Bulkhead` on start method.
Having the fault tolerance capability is probably fine but why not other options (e.g. `@Retry`)? Why we want to work with only 10 parallel requests - what is the reason? Are we fine in adding one more dependency to the mix for the LRA processing?

mmusgrov replied:
Because we want to protect the coordinator against overload. This PR is to address a particular issue I found during my stress testing of the coordinator (the issue really needs to be a blocker). Throttling start requests will reduce the number of active LRAs that the coordinator manages. So applying a bulkhead to start will reduce the total number of parallel LRAs to a manageable number and the other coordinator requests will all apply to this reduced number of LRAs.

I chose the default because, well it is the default. The user may then configure a value that is appropriate to his installation (using the usual MP FT config mechanism).

I didn't need any of the other FT annotations because the intent of the PR is to protect the coordinator. In the PR description I did explain why we don't use most of the FT features:

> There is a lot we can do in this area [fault resilience] and we have a few JIRAs for tackling the various requirements. But this PR is a basic minimum to reduce the number of errors caused by overloading the system in which LRA operates.

xstefank replied:
Generally, this PR looks good. But I just want to point out that `@Bulkhead` is only reducing concurrent executions of the method thus at max 10 thread can run `start` at the same time but that doesn't correspond to the reduction of the number of active LRAs that the coordinator manages. The errornous state can still be reached but it would take a few retries on the /start calls.

ochaloup replied:
@mmusgrov sure, I was not aware about the problem from the stress testing. Then this makes sense. Thanks.

mmusgrov replied:
I ran a local test to ensure that the Bulkhead was being respected.

ochaloup commented on rts/lra/coordinator/pom.xml:
```diff
@@ -28,6 +28,11 @@
         </dependency>
+        <dependency>
+            <groupId>org.eclipse.microprofile.fault-tolerance</groupId>
+            <artifactId>microprofile-fault-tolerance-api</artifactId>
```
Is the compile scope considered as the right one for this?

mmusgrov replied:
That's a good point, I cut and pasted the dependency from maven central. But you are right in querying this, I would have expected we'd need to be pulling in the smallrye implementation of FT. I will investigate further and update my PR depending upon what I find.

I did find that after I made the changes in this PR that I resolved the errors produced by the coordinator under heavy load (something like 5000 out of 5000 successful LRAs against two remote participants, it did take a couple of recovery passes for them all to finish okay). So I assumed my fix was good and did not drill down any further into the bulkhead.

ochaloup replied:
I think the API will be bundled to the final jar file and the quarkus/thorntail is then taking the smallrye implementation into consideration and start it.
The trouble could be with WFLY maybe. In fact when I was looking at the `pom.xml` I'm not sure, from this perspective, is good to have the compile scope for `microprofile-lra-api` either. The deployment should be using what's provided by app runtime. I hope we can consider this as a nuance that could be investigated when working on WFLY extension.

mmusgrov replied:
Well I made some changes to how we consume the dependency although I didn't quite follow your last comment so you might want to check my latest push.

ochaloup replied:
My point is the best described in way that I think the dependency should be in scope `provided`. The app runtime then loads the api library. It won't be the application that defines it.

mmusgrov replied:
+1

ochaloup reviewed (commented):
I have few questions but in general it's probably fine.

mmusgrov commented:
@xstefank yes agreed and this will still take the load off the coordinator.
I think it would be useful to still cap the total number of active LRAs, can you recommend any other FT annotations that would help us in this respect?

ochaloup commented:
@mmusgrov I wonder about the idea of capping the number of active LRAs. Narayana does not do any similar thing for JTA or JTS, I think.

ochaloup reviewed (approved):
Thanks for the elaboration on my comments. There could be some further improvements but nothing big for not approving.

mmusgrov commented:
> @mmusgrov I wonder about the idea of capping the number of active LRAs. Narayana does not do any similar thing for JTA or JTS, I think.

The cap is a config property for users to apply only if their environment is known be constrained, AMS for example, (and we can make the default as large or as small as we like).

We don't do this with JTA or JTS because they run in app server environments whose resource needs are generally well known. To date we haven't provided any transaction products that by design are meant to run in a loosely coupled distributed system (XTS is the closest but that still runs with the app server as a primary architectural piece).

And MP-LRA is a MP component and use of other MP components is encouraged in order to achieve the use cases for which MicroProfile was created. MP-FT is designed for those environments where we expect users to be running LRA.

xstefank commented:
@mmusgrov FT isn't capable of such a thing up to my knowledge. The only thing that comes to mind is to cut the size of activeLRAs map in LRAService. Not sure if that is the correct solution.

Also off-topic, can we then move the configuration to MP config? I would like to work on that.

mmusgrov commented:
> @mmusgrov FT isn't capable of such a thing up to my knowledge. The only thing that comes to mind is to cut the size of activeLRAs map in LRAService. Not sure if that is the correct solution.

Okay, then let's stay with just `@Bulkhead` for now since that seemed to resolve the immediate problem I found in my stress testing.

> 
> Also off-topic, can we then move the configuration to MP config? I would like to work on that.

Would doing this in the narayana code base interfere with what we are doing in the Quarkus and WildFly integrations. If not then, yes, I agree that leveraging MP config would be good.

xstefank commented:
OK. Config shouldn't interfere. SR projects often come with their own config properties. quarkus prefix should be restricted to quarkus specifics I believe. In every case, please go ahead and merge this PR, I'll create a new JBTM to track config.

mmusgrov commented:
@xstefank and @ochaloup I made a number of changes since you both approved it so you may want to rescan the PR to verify that I addressed your comments correctly.

ochaloup commented on rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java:
```diff
@@ -87,8 +87,7 @@
     private static final Pattern START_END_QUOTES_PATTERN = Pattern.compile("^\"|\"$");
     private static final long DEFAULT_TIMEOUT_MILLIS = 0L;
-    // i18nMessageCode corresponding to lraI18NLogger#warn_LRAStatusInDoubt
-    private static final int i18nMessageCode = 25145;
```
I like this change ;-)

ochaloup reviewed (approved):
I think I'm fine with the current state of the PR.
